include/subsystems/odometry/odometry_nwheel.h

```diff
@@ -0,0 +1,255 @@
+      this->ang_accel_deg = ang_accel_local;
+    }
+
+    old_heading = angle;
+
+    return current_pos;
+  }
+
+  /**
+   * Resets the position and rotational data to the input.
+   */
+  void set_position(const pose_t &newpos) override {
+    mut.lock();
+    rotation_offset = newpos.rot - (current_pos.rot - rotation_offset);
+    mut.unlock();
+
+    OdometryBase::set_position(newpos);
+  }
+
+private:
+  /**
+   * Applies a twist (pose delta) to a pose by including first order dynamics of heading.
+   * Can be thought of as applying a twist as following an arc rather than a straight line.
+   *
+   * https://file.tavsys.net/control/controls-engineering-in-frc.pdf#section.10.2
+   *
+   * @param old_pose  The pose to which the twist will be applied
+   * @param twist     The twist, represents a pose delta
+   */
+  pose_t pose_exponential(const Eigen::Vector3d old_pose, const Eigen::Vector3d twist) {
+    double dtheta = twist(2);
+
+    double sinTheta = sin(dtheta);
+    double cosTheta = cos(dtheta);
+
+    double sinOldTheta = sin(old_pose(2));
+    double cosOldTheta = cos(old_pose(2));
+
+    Eigen::Matrix3d rotation{{cosOldTheta, -sinOldTheta, 0}, {sinOldTheta, cosOldTheta, 0}, {0, 0, 1}};
+    Eigen::Matrix3d integrated_rotation;
+
+    // when the angle change is very small, we use a taylor series to approximate
+    if (abs(dtheta) < 1e-9) {
+      integrated_rotation.row(0) << 1.0 - ((dtheta * dtheta) / 6.0), -(dtheta / 2.0), 0;
+      integrated_rotation.row(1) << (dtheta / 2.0), 1.0 - ((dtheta * dtheta) / 6.0), 0;
+      integrated_rotation.row(2) << 0, 0, 1;
+    } else {
+      integrated_rotation.row(0) << sinTheta / dtheta, (cosTheta - 1.0) / dtheta, 0;
+      integrated_rotation.row(1) << (1 - cosTheta) / dtheta, sinTheta / dtheta, 0;
+      integrated_rotation.row(2) << 0, 0, 1;
+    }
+
+    Eigen::Vector3d global_pose_delta = rotation * integrated_rotation * twist;
+
+    pose_t newPose;
+
+    newPose.x = old_pose(0) + global_pose_delta(0);
+    newPose.y = old_pose(1) + global_pose_delta(1);
+    newPose.rot = wrap_angle_rad(old_pose(2) + global_pose_delta(2));
+
+    return newPose;
+  }
+
+  /**
+   * Calculation method for the robot's new position using the change in encoders, and the old pose, the wheel
+   * configurations are stored as class members.
+   *
+   * @param radian_deltas vector containing the change of angle of each wheel (radians)
+   * @param old_pose The robot's previous position (x, y, rot)
+   * @return The robot's new position (x, y, rot)
+   */
+  pose_t calculate_new_pos(Eigen::Vector<double, WHEELS> radian_deltas, pose_t old_pose) {
+    // Mr T = E -> Mr^{+} E = T
+    Eigen::Vector3d pose_delta = transfer_matrix_pseudoinverse * (radian_deltas.asDiagonal() * wheel_radii);
+    Eigen::Vector3d old_pose_vector{old_pose.x, old_pose.y, old_pose.rot};
+
+    // we achieve better performance by using the imu for rotation directly
+    pose_delta(2) = angle - old_heading;
```

Comment:
not that it would happen often, but just for the flexibility, if its not too hard can you put in a `if imu!=nullptr` use imu else use wheels. probably won't happen too often but itd be nice to have the option

Reply:
and then just update the comment to say that passing in null as imu is valid and will just use the wheels

Reply:
okay this also works